### PR TITLE
bpo-30622: Change NPN detection

### PR DESCRIPTION
**Pull request by Sp1l**
This aligns the way NPN support is detected in the code. It now works the same way that ALPN support is detected and used.

https://bugs.python.org/issue30622

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by tiran** [comment]
Thanks! I have another NPN related fix in #5253. Are you ok if I merge your improvements with mine and give you credit for your work?

**Comment by Sp1l** [comment]
@tiran No problemo. Let me check #5253 
You're OK with the general direction on the NPN feature detection?

**Comment by Sp1l** [comment]
@tiran Merged your changes into this PR.

CLA has been signed based on my bugs.python.org id 'spil'

**Comment by tiran** [comment]
Awesome! Thanks! 👍 
 
* The-knights-who-say-ni don't see your CLI yet. It may take a day or two until it is processed by a human being.
* You have to regenerate the argument clinic, please run ``make clinic`` and push the modified header file.
* I'm currently travelling. Your fix may not land in 3.7.0-beta 1.

File: Modules/_ssl.c [hunk 1]
```diff
@@ -157,6 +157,25 @@ static void _PySSLFixErrno(void) {
 # define HAVE_ALPN
 #endif
 
+/* NPN support defines changed in OpenSSL 1.1
+ * LibreSSL 2.6.1 no longer supports NPN but does not set the
+ * designated OPENSSL_NO_NEXTPROTONEG feature flag. See upstream issue
+ * https://github.com/libressl-portable/portable/issues/368
+ */
+#ifdef OPENSSL_VERSION_1_1
+# ifdef OPENSSL_NO_NEXTPROTONEG
```
Comment (tiran) [inline Modules/_ssl.c]: I think it is safe to simplify the check even further. NPN is available if ``TLSEXT_TYPE_next_proto_neg`` and ``OPENSSL_NO_NEXTPROTONEG`` is not defined.

```
#ifdef OPENSSL_NO_NEXTPROTONEG
#  define HAVE_NPN 0
#elif defined(TLSEXT_TYPE_next_proto_neg)
#  define HAVE_NPN 1
#else
#  define HAVE_NPN 0
# endif
```

**Review by tiran — CHANGES_REQUESTED** [review]
Please rebase your patch and simplify the check as explained.

**Comment by tiran** [comment]
I have resolved the merge conflict through GH ui. Tests are still failing because the clinic file is outdated.

**Comment by tiran** [comment]
Somehow I messed up your PR with Github's UI. Since 3.7.0b2 is immanent, I created PR #5859 and gave you credit as co-author.

**Pull request closed without merging**